lib/rules/no-extra-parens.js

```diff
@@ -386,6 +386,31 @@ module.exports = {
             return node && (node.type === "Identifier" || node.type === "MemberExpression");
         }
 
+        /**
+         * Checks if a node is fixable.
+         * A node is fixable if removing a single pair of surrounding parentheses does not turn it
+         * into a directive.
+         * Almost all nodes are fixable, except if all of the following conditions are met:
+         * The node is a string Literal
+         * It has a single pair of parentheses
+         * It is the only child of an ExpressionStatement
+         * It immediately follows a (possibly empty) directive prologue
+         * @param {ASTNode} node The node to evaluate.
+         * @returns {boolean} Whether or not the node is fixable.
+         * @private
+         */
+        function isFixable(node) {
+
+            // if it's not a string literal it can be autofixed
+            if (node.type !== "Literal" || typeof node.value !== "string") {
+                return true;
+            }
+            if (isParenthesisedTwice(node)) {
+                return true;
+            }
+            return !astUtils.isExpressionInOrJustAfterDirectivePrologue(sourceCode, node.parent);
```

Comment:
I thought we'd just check if the parent is ExpressionStatement as we don't know what fixes other rules are applying in the code that precedes it.

For example:

```js
/* eslint no-extra-parens: "error" */
/* eslint prettier/prettier: "error" */

; //
("use strict");

```

after `--fix`, `"use strict"` still ends up being a directive:

```js
/* eslint no-extra-parens: "error" */
/* eslint prettier/prettier: "error" */

//
"use strict";

```

Reply:
Well, if compatibility with Prettier is a concern then we could just check the parent node as you suggest. I don't know how we could stop third-party rules from interfering if another autofix is applied to the `ExpressionStatement` node itself or to its children rather than to a sibling node (e.g. `("use " + "strict");` → `("use strict");`), but from what I can tell, that is not what Prettier does.

In this case we should also change the behavior of the `quotes` rule, because fox example

```js
/* eslint quotes: "error" */
/* eslint prettier/prettier: "error" */

; //
`use strict`;
```

autofixes to

```js
/* eslint quotes: "error" */
/* eslint prettier/prettier: "error" */

//
"use strict";
```

Reply:
> Well, if compatibility with Prettier is a concern then we could just check the parent node as you suggest.

Not specifically with Prettier, any custom or third-party rules.

> I don't know how we could stop third-party rules from interfering if another autofix is applied to the `ExpressionStatement` node itself or to its children rather than to a sibling node (e.g. `("use " + "strict");` → `("use strict");`),

That shouldn't cause problems because the ranges will be overlapping so ESLint would apply only one of the fixes:

https://eslint.org/docs/latest/extend/custom-rules#conflicting-fixes

I think that non-directive ExpressionStatement > Literal is a very edge case, and it's a meaningless code that should be manually fixed anyway, so I'd vote for the safest solution which is also the simplest for us. @nzakas what do you think?

> ```js
> /* eslint quotes: "error" */
> /* eslint prettier/prettier: "error" */
> 
> ; //
> `use strict`;
> ```

Yes, I don't think this should be autofixed either.

Reply:
> That shouldn't cause problems because the ranges will be overlapping so ESLint would apply only one of the fixes:
> 
> https://eslint.org/docs/latest/extend/custom-rules#conflicting-fixes

Ah, understood. I didn't realize that removing the parentheses is handled as one single fix by the linter. Thanks for the explanation @mdjermanovic.